### PR TITLE
Finalization service

### PR DESCRIPTION
**Pull request by chrisseaton**

### CODE REVIEW

File: src/main/java/org/truffleruby/core/FinalizationService.java [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+        }
+    }
+
+    private void createFinalizationThread() {
+        finalizerThread = context.getThreadManager().createRubyThread("finalizer");
+        context.send(finalizerThread, "internal_thread_initialize", null);
+
+        ThreadManager.initialize(finalizerThread, context, null, "finalizer", () -> {
+            while (true) {
+                final FinalizerReference finalizerReference = (FinalizerReference) context.getThreadManager().runUntilResult(null, () -> finalizerQueue.remove());
```
Comment (eregon) [inline src/main/java/org/truffleruby/core/FinalizationService.java]: This line is really long, could it be cut in half somehow?

File: src/main/java/org/truffleruby/core/objectspace/ObjectSpaceManager.java [hunk 2]
```diff
@@ -101,75 +80,26 @@ public void clearFinalizers() {
...
+        }, () -> {
+            if (callable instanceof DynamicObject) {
+                return Stream.of((DynamicObject) callable);
+            } else {
+                return Stream.empty();
             }
-        } catch (RaiseException e) {
-            // MRI seems to silently ignore exceptions in finalizers
-        }
+        });
```
Comment (eregon) [inline src/main/java/org/truffleruby/core/objectspace/ObjectSpaceManager.java]: Would it be simpler to just have a List<DynamicObject> for the roots?

File: src/main/java/org/truffleruby/core/FinalizationService.java [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+        public FinalizerReference(Object object, ReferenceQueue<? super Object> queue) {
+            super(object, queue);
+        }
+
+        public void addFinalizer(Class<?> owner, Runnable action, Supplier<Stream<DynamicObject>> roots) {
+            finalizers.add(new Finalizer(owner, action, roots));
+        }
+
+        public void removeFinalizers(Class<?> owner) {
+            finalizers.removeIf((f) -> f.getOwner() == owner);
```
Comment (eregon) [inline src/main/java/org/truffleruby/core/FinalizationService.java]: FWIW lambdas with a single argument can be written without the extra ():
```java
finalizers.removeIf(f -> f.getOwner() == owner);
```
I find it more readable.

File: test/truffle/integration/finalizers.sh [hunk 4]
```diff
@@ -2,4 +2,6 @@
 
 source test/truffle/common.sh.inc
 
-jt ruby test/truffle/integration/finalizers/define_finalizer.rb
+for [f in test/truffle/integration/finalizers/*.rb]; do
```
Comment (eregon) [inline test/truffle/integration/finalizers.sh]: What is this syntax with the `[`? Are they necessary?

Comment (eregon) [inline test/truffle/integration/finalizers.sh]: The CI does not like it so I guess a typo.

File: src/main/java/org/truffleruby/core/FinalizationService.java (hunk 3 above)
Comment (chrisseaton) [inline src/main/java/org/truffleruby/core/FinalizationService.java]: Fixed.

File: src/main/java/org/truffleruby/core/objectspace/ObjectSpaceManager.java (hunk 2 above)
Comment (chrisseaton) [inline src/main/java/org/truffleruby/core/objectspace/ObjectSpaceManager.java]: Fixed.

File: test/truffle/integration/finalizers.sh (hunk 4 above)
Comment (chrisseaton) [inline test/truffle/integration/finalizers.sh]: Fixed.

**Pull request merged**